File: server/src/routes/getRandomWallet.js
// IMPORTS
const axios = require('axios');
require('dotenv').config();

// UTILITIES
const {
  cloudinary,
  imagekit,
  DEFAULT_CLOUDINARY_IMG,
  DEFAULT_IMAGEKIT_IMG,
} = require('../utils/image.js');
const changeIpfsUrl = require('../utils/changeIpfsUrl.js');
const resolveDomain = require('../utils/resolve.js');

const getRandomWallet = async (req, res) => {
  const now = Math.floor(Date.now());

  // Moralis getDateToBlock
  // get latest Ethereum block
  let response = await axios
    .get(`${process.env.MORALIS_API_URL}/dateToBlock?chain=eth&date=${now}`, {
      headers: {
        accept: 'application/json',
        'X-API-KEY': process.env.MORALIS_API_KEY,
      },
    })
    .catch((err) => {
      console.log(err);
    });

  const latestBlock = response.data.block;

  // get a list of NFT transactions
  // Moralis GetNFTTransfersByBlock
  response = await axios
    .get(
      `${process.env.MORALIS_API_URL}/block/${latestBlock}/nft/transfers?chain=eth&limit=250`,
      {
        headers: {
          accept: 'application/json',
          'X-API-KEY': process.env.MORALIS_API_KEY,
        },
      }
    )
    .catch((err) => {
      console.log(err);
    });

  const transactions = response.data.result;

  const rand = Math.floor(Math.random() * transactions.length);

  try {
    if (transactions[rand].to_address) {
      res.send(transactions[rand].to_address);
    } else {
      res.send(transactions[rand].from_address);
    }
  } catch (err) {
    console.log(err);
  }
};

module.exports = getRandomWallet;

File: go/internal/routes/getRandomWallet.go
package routes

import (
	"api/pkg/request"
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"strconv"
	"time"
)


func GetRandomWallet(w http.ResponseWriter, r *http.Request) {

	type Block struct {
		Date string `json:"date"`
		Block int `json:"block"`
		Timestamp int `json:"timestamp"`
	}
	
	type Result struct {
		Block_Number string `json:"block_number"`
		Block_Timestamp string `json:"block_timestamp"`
		Block_Hash string `json:"block_hash"`
		Transaction_Hash string `json:"transaction_hash"`
		Transaction_Index int `json:"transaction_index"`
		Log_Index int `json:"log_index"`
		Value string `json:"value,omitempty"`
		Contract_Type string `json:"contract_type"`
		Transaction_Type string `json:"transaction_type"`
		Token_Address string `json:"token_address"`
		Token_Id string `json:"token_id"`
		From_Address string
		To_Address string `json:"to_address"`
	}
	
	type Data struct {
		Result []Result `json:"result"`
		Total int `json:"total"`
	}

	// Get current timestamp to get latest block and convert to string
	now := time.Now().UnixMilli()
	timestamp := strconv.FormatInt(now, 10)

	// Moralis getDateToBlock https://docs.moralis.io/moralis-dapp/web3-sdk/native#getdatetoblock
	response, err := request.APIRequest(`/dateToBlock?chain=eth&date=` + timestamp)
	if err != nil {
		fmt.Println("API Request Error", err)
		return
	}

	var block Block

	err = json.Unmarshal([]byte(response), &block)
	if err != nil {
		fmt.Println("Couldn't unmarshal: ", err)
		return
	}

	latestBlock := strconv.Itoa(block.Block) 


	// Moralis GetNFTTransfersByBlock https://github.com/nft-api/nft-api#GetNFTTransfersByBlock
	response, err = request.APIRequest(`/block/` + latestBlock + `/nft/transfers?chain=eth&limit=250`)
	if err != nil {
		fmt.Println("API Request Error", err)
		return
	}

	var data Data

	err = json.Unmarshal([]byte(response), &data)
	if err != nil {
		fmt.Println("Couldn't unmarshal: ", err)
		return
	}

	if data.Total > 0 {
		rand := rand.Intn(data.Total) // generate random number from 1 to no. of transfers

		w.WriteHeader(http.StatusOK)

		if len(data.Result[rand].To_Address) != 0 {
			if data.Result[rand].To_Address != "" && data.Result[rand].To_Address != "0x0000000000000000000000000000000000000000" && data.Result[rand].To_Address != "0x000000000000000000000000000000000000dead" {
				fmt.Fprintf(w, "%s", data.Result[rand].To_Address)
			}
		} else if len(data.Result[rand].From_Address) != 0 {
			if data.Result[rand].From_Address != "" && data.Result[rand].From_Address != "0x0000000000000000000000000000000000000000" && data.Result[rand].From_Address != "0x000000000000000000000000000000000000dead" {
				fmt.Fprintf(w, "%s", data.Result[rand].From_Address)
			}
		}
	}
}